File: java/creational/ObjectPoolPattern.java
package creational;

import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Object Pool Pattern
 * Reuses objects that are expensive to create.
 */
public class ObjectPoolPattern {
    
    // Pooled object
    static class DatabaseConnection {
        private static int counter = 0;
        private final int id;
        private boolean inUse;
        
        public DatabaseConnection() {
            this.id = ++counter;
            System.out.println("Creating new connection #" + id);
        }
        
        public void connect() {
            System.out.println("Connection #" + id + " - Executing query...");
        }
        
        public void disconnect() {
            System.out.println("Connection #" + id + " - Query completed");
        }
        
        public boolean isInUse() {
            return inUse;
        }
        
        public void setInUse(boolean inUse) {
            this.inUse = inUse;
        }
        
        public int getId() {
            return id;
        }
    }
    
    // Object Pool
    static class ConnectionPool {
        private final int maxPoolSize;
        private final ConcurrentLinkedQueue<DatabaseConnection> availableConnections;
        private final ConcurrentLinkedQueue<DatabaseConnection> inUseConnections;
        
        public ConnectionPool(int maxPoolSize) {
            this.maxPoolSize = maxPoolSize;
            this.availableConnections = new ConcurrentLinkedQueue<>();
            this.inUseConnections = new ConcurrentLinkedQueue<>();
            
            // Pre-populate pool
            for (int i = 0; i < maxPoolSize / 2; i++) {
                availableConnections.add(new DatabaseConnection());
            }
        }
        
        public synchronized DatabaseConnection acquireConnection() {
            DatabaseConnection connection = availableConnections.poll();
            
            if (connection == null) {
                if (inUseConnections.size() < maxPoolSize) {
                    connection = new DatabaseConnection();
                } else {
                    System.out.println("⚠️ Pool exhausted! Waiting for available connection...");
                    return null;
                }
            }
            
            connection.setInUse(true);
            inUseConnections.add(connection);
            System.out.println("✓ Acquired connection #" + connection.getId() + 
                             " (Available: " + availableConnections.size() + 
                             ", In use: " + inUseConnections.size() + ")");
            return connection;
        }
        
        public synchronized void releaseConnection(DatabaseConnection connection) {
            if (connection != null && inUseConnections.remove(connection)) {
                connection.setInUse(false);
                availableConnections.add(connection);
                System.out.println("✓ Released connection #" + connection.getId() + 
                                 " (Available: " + availableConnections.size() + 
                                 ", In use: " + inUseConnections.size() + ")");
            }
        }
        
        public void shutdown() {
            availableConnections.clear();
            inUseConnections.clear();
            System.out.println("Pool shutdown complete");
        }
    }
    
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Object Pool Pattern Demo ===\n");
        
        ConnectionPool pool = new ConnectionPool(3);
        
        System.out.println("\n--- Acquiring connections ---");
        DatabaseConnection conn1 = pool.acquireConnection();
        if (conn1 != null) conn1.connect();
        
        DatabaseConnection conn2 = pool.acquireConnection();
        if (conn2 != null) conn2.connect();
        
        DatabaseConnection conn3 = pool.acquireConnection();
        if (conn3 != null) conn3.connect();
        
        // Try to acquire when pool is exhausted
        System.out.println("\n--- Pool exhausted scenario ---");
        DatabaseConnection conn4 = pool.acquireConnection();
        
        // Release and reuse
        System.out.println("\n--- Releasing and reusing connections ---");
        if (conn1 != null) {
            conn1.disconnect();
            pool.releaseConnection(conn1);
        }
        
        DatabaseConnection conn5 = pool.acquireConnection();
        if (conn5 != null) {
            conn5.connect();
            System.out.println("Reused connection #" + conn5.getId());
        }
        
        // Cleanup
        System.out.println("\n--- Cleanup ---");
        if (conn2 != null) pool.releaseConnection(conn2);
        if (conn3 != null) pool.releaseConnection(conn3);
        if (conn5 != null) pool.releaseConnection(conn5);
        
        pool.shutdown();
    }
}

File: python/creational/object_pool_pattern.py
"""
Object Pool Pattern
Purpose: Reuses objects that are expensive to create
Use Case: Database connections, thread pools, network connections
"""

from typing import Generic, TypeVar, List, Optional
import threading
import time


T = TypeVar('T')


class ObjectPool(Generic[T]):
    def __init__(self, factory, max_size: int = 10):
        self._factory = factory
        self._max_size = max_size
        self._available: List[T] = []
        self._in_use: List[T] = []
        self._lock = threading.Lock()
    
    def acquire(self) -> Optional[T]:
        with self._lock:
            if self._available:
                obj = self._available.pop()
                self._in_use.append(obj)
                return obj
            elif len(self._in_use) < self._max_size:
                obj = self._factory()
                self._in_use.append(obj)
                return obj
            return None
    
    def release(self, obj: T):
        with self._lock:
            if obj in self._in_use:
                self._in_use.remove(obj)
                self._available.append(obj)
    
    def __str__(self):
        return f"Pool(available={len(self._available)}, in_use={len(self._in_use)})"


class DatabaseConnection:
    _counter = 0
    
    def __init__(self):
        DatabaseConnection._counter += 1
        self.id = DatabaseConnection._counter
        print(f"  [CREATED] Connection #{self.id}")
    
    def query(self, sql: str):
        return f"Connection #{self.id} executed: {sql}"
    
    def __str__(self):
        return f"Connection #{self.id}"


if __name__ == "__main__":
    print("=== Object Pool Pattern Demo ===\n")
    
    pool = ObjectPool(DatabaseConnection, max_size=3)
    
    print("Acquiring connections...")
    conn1 = pool.acquire()
    conn2 = pool.acquire()
    print(f"{pool}\n")
    
    print(f"{conn1.query('SELECT * FROM users')}")
    
    print("\nReleasing connection...")
    pool.release(conn1)
    print(f"{pool}\n")
    
    print("Reusing released connection...")
    conn3 = pool.acquire()  # Reuses conn1
    print(f"{pool}")
